### DISCUSSION TITLE
Edit hosts file

### DISCUSSION THREAD

**Discussion by mdeppe42** [nextcloud/all-in-one#2395] [Questions] [answered]
Hi,
is there any possibility to permanently edit the hosts file of the nextcloud-aio-nextcloud container?
I need a line added to this file. I did it by logging into the docker contailer using docker exec, and then editing the /etc/hosts file. It's working fine, but I guess the change will be lost when a container update is performed?

I already tried adding the --add-host parameter to the docker run command of the mastercontainer, but this doesn't affect the hosts file of the nexcloud container.

**Comment by szaimen** [comment]
Hi, why would you need such a possibility?

**Reply by mdeppe42** [reply]
I have a mail server (docker-mailserver) running on the same host in another container. The mail server container is added to the nextcloud-aio docker network. Inside this network the mail server can be accessed by the host name "mailserver". Setting the SMTP server address in Nextcloud to "mailserver" or the IP address inside the docker network doesn't work because of the certificate, which is issued to the domain name "mail.xxxxx.tld".
I manually edited the hosts file of the nextcloud-aio-nextcloud container and added the line
<IP of mail server inside the docker network> mail.xxxxx.tld
and this works. But of course I need this to be a permanent solution.

(Without editing the hosts file, mail.xxxxx.tld would resolve to the public IP of my server, but the SMTP port is not accessible)

**Comment by szaimen** [comment]
Hi, you can easily set up a local dns server by using one of the following guides: https://github.com/nextcloud/all-in-one#how-can-i-access-nextcloud-locally and put the config in there as custom dns record for that specific pointer.

**Reply by MrRinkana** [reply]
@szaimen  I'm not sure this answers the question, the nextcloud-aio-nextcloud container needs 
`172.18.0.1 mail.example.com` but the rest of the (non docker) network cannot have that as DNS, as mail.example.com needs to point to the hosts external ip (not the internal docker network). In essence one would need a DNS server for **only** specifically the nextcloud-aio-nextcloud container and nothing else.

That's why people are asking for the possibility of adding lines to the nextcloud-aio-nextclouds **own** /etc/hosts

**Comment by renne** [comment]
@szaimen For reliability reasons (e.g. dns resolver fail) it would be helpful to add `extra_hosts` to the containers.
In our case we use the Nginx-Proxy-Manager community add-on. If the external dns resolver fails Nginx-Proxy-Manager should still be able to resolve the IP adresses of Nextcloud-AIO and other docker containers correctly  (e.g. `- docker.example.com:host-gateway`).

**Reply by xstasi** [reply]
+1 if anyone is still reading this, I'd love to have a solution to have `extra_hosts` in the containers.

**Reply by furynick** [reply]
agreed too, my situation need it also.
I have to manually patch hosts files to workaround curl error at container startup.

**Reply by dylanmeysmans** [reply]
@szaimen I would also like to have this option.

**Reply by MrRinkana** [reply]
This is also useful if the vps running does not allow for outgoing SMTP but you have a email server on the same host using an approved relay

**Comment by Johar-tumen** [comment]
At the moment, the only working solution is to add a cron task to replace the hosts file in specific containers. These containers are usually located in /var/lib/docker/containers/container_UUID/. Inside the folder, you can find the hostname file, which allows you to identify the container and replace the hosts file without restarting the container.

**Reply by Johar-tumen** [reply]
But it would definitely be better if there was a setting - extra-hosts for some containers. In my case, the server is located behind a NAT router, and when collabora-office wants to open something at my_domain.com, for example, it goes to the router's IP, which redirects the request to another web interface. This causes an error.

**Reply by Vyeche** [reply]
Good idea, the CRON job host file worked for me. Hopefully the add extra hosts.

**Comment by szaimen** [comment]
Hi, I am working on adding some better instructions to the readme: https://github.com/nextcloud/all-in-one/pull/7607

**Reply by szaimen** [reply, marked as answer]
See https://github.com/nextcloud/all-in-one?tab=readme-ov-file#how-to-overwrite-the-local-dns-resolution-for-some-domains-or-add-extra-hosts-to-the-containers
